File: user-service/app/services/service_client.py
"""
Service Client for User Service
Handles inter-service communication using Consul service discovery
"""
import aiohttp
import asyncio
import json
from typing import Dict, Any, Optional, List
from app.utils.consul import consul_service
from app.core.config import settings
from app.core.logger import logger


class ServiceClient:
    """HTTP client for inter-service communication with service discovery"""
    
    def __init__(self):
        self.timeout = aiohttp.ClientTimeout(total=30)
        self.headers = {
            'Content-Type': 'application/json',
            'User-Agent': 'user-service/1.0'
        }
    
    async def _make_service_request(
        self,
        service_name: str,
        endpoint: str,
        method: str = 'GET',
        data: Optional[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        Make HTTP request to a discovered service
        
        Args:
            service_name: Name of the service to call
            endpoint: API endpoint to call
            method: HTTP method (GET, POST, etc.)
            data: Request body data
            
        Returns:
            Response data as dictionary
            
        Raises:
            Exception: If service discovery or request fails
        """
        try:
            # Discover the service
            service_info = consul_service.get_service(service_name)
            if not service_info:
                raise Exception(f"Service {service_name} not found in Consul")
            
            # Build URL
            url = f"http://{service_info['address']}:{service_info['port']}{endpoint}"
            logger.info(f"Making {method} request to {service_name}: {url}")
            
            # Prepare request data
            json_data = None
            if data and method.upper() in ['POST', 'PUT', 'PATCH']:
                json_data = data
            
            # Make the request
            async with aiohttp.ClientSession(timeout=self.timeout) as session:
                async with session.request(
                    method.upper(),
                    url,
                    headers=self.headers,
                    json=json_data
                ) as response:
                    
                    if response.status >= 400:
                        error_text = await response.text()
                        raise Exception(f"HTTP {response.status}: {error_text}")
                    
                    # Try to parse JSON response
                    try:
                        return await response.json()
                    except:
                        # If not JSON, return text
                        text_response = await response.text()
                        return {"response": text_response}
                    
        except Exception as e:
            logger.error(f"Service call to {service_name} failed: {e}")
            raise
    
    async def validate_auth_token(self, token: str) -> Dict[str, Any]:
        """
        Validate authentication token with login service
        
        Args:
            token: JWT token to validate
            
        Returns:
            Validation result with user info
        """
        try:
            logger.info("Validating auth token with login-service")
            
            result = await self._make_service_request(
                'login-service',
                '/auth/validate',
                'POST',
                {'token': token}
            )
            
            return {
                'valid': result.get('valid', False),
                'user_id': result.get('user_id'),
                'email': result.get('email')
            }
            
        except Exception as e:
            logger.error(f"Auth token validation failed: {e}")
            return {'valid': False}
    
    async def notify_order_service(self, user_id: str, event: str, data: Dict[str, Any]) -> bool:
        """
        Notify order service about user-related events
        
        Args:
            user_id: User ID
            event: Event type (user_updated, user_deleted, etc.)
            data: Event data
            
        Returns:
            True if notification successful, False otherwise
        """
        try:
            logger.info(f"Notifying order-service about {event} for user {user_id}")
            
            await self._make_service_request(
                'order-service',
                '/notifications/user-event',
                'POST',
                {
                    'user_id': user_id,
                    'event': event,
                    'data': data,
                    'timestamp': settings.get_current_timestamp()
                }
            )
            
            logger.info("Order service notification sent successfully")
            return True
            
        except Exception as e:
            # Don't fail the operation if notification fails
            logger.warning(f"Failed to notify order service: {e}")
            return False
    
    async def check_service_health(self, service_name: str) -> bool:
        """
        Check if a service is healthy
        
        Args:
            service_name: Name of the service to check
            
        Returns:
            True if service is healthy, False otherwise
        """
        try:
            await self._make_service_request(service_name, '/health', 'GET')
            return True
        except:
            return False
    
    async def get_available_services(self) -> List[str]:
        """
        Get list of available and healthy services
        
        Returns:
            List of service names that are available
        """
        try:
            services = ['login-service', 'order-service']
            available_services = []
            
            for service in services:
                is_healthy = await self.check_service_health(service)
                if is_healthy:
                    available_services.append(service)
            
            logger.info(f"Available services: {available_services}")
            return available_services
            
        except Exception as e:
            logger.error(f"Failed to check available services: {e}")
            return []
    
    async def get_user_orders(self, user_id: str) -> List[Dict[str, Any]]:
        """
        Get user's orders from order service
        
        Args:
            user_id: User ID
            
        Returns:
            List of user's orders
        """
        try:
            logger.info(f"Getting orders for user {user_id} from order-service")
            
            result = await self._make_service_request(
                'order-service',
                f'/order/user/{user_id}',
                'GET'
            )
            
            return result.get('orders', [])
            
        except Exception as e:
            logger.error(f"Failed to get user orders: {e}")
            return []


# Global service client instance
service_client = ServiceClient()

File: order-service/src/services/serviceClient.ts
/**
 * Service Communication Module for Order Service
 * Handles inter-service communication using Consul service discovery
 */
import * as http from 'http';
import * as https from 'https';
import consulService from './consulService';

/**
 * HTTP Client for inter-service communication
 */
class ServiceClient {
  private baseHeaders: Record<string, string>;

  constructor() {
    this.baseHeaders = {
      'Content-Type': 'application/json',
      'User-Agent': 'order-service/1.0'
    };
  }

  /**
   * Make HTTP request to a discovered service using Node.js built-in modules
   */
  private async makeServiceRequest(
    serviceName: string,
    endpoint: string,
    method: string = 'GET',
    body?: any
  ): Promise<any> {
    return new Promise(async (resolve, reject) => {
      try {
        // Discover the service
        const serviceInfo = await consulService.discoverService(serviceName);
        console.log(`📞 Calling ${serviceName} at ${serviceInfo.url}${endpoint}`);

        // Parse URL
        const url = new URL(`${serviceInfo.url}${endpoint}`);
        const isHttps = url.protocol === 'https:';
        const client = isHttps ? https : http;

        // Prepare request data
        let postData = '';
        if (body && (method === 'POST' || method === 'PUT' || method === 'PATCH')) {
          postData = JSON.stringify(body);
        }

        // Request options
        const options = {
          hostname: url.hostname,
          port: url.port || (isHttps ? 443 : 80),
          path: url.pathname + url.search,
          method: method.toUpperCase(),
          headers: {
            ...this.baseHeaders,
            'Content-Length': Buffer.byteLength(postData)
          }
        };

        // Make the request
        const req = client.request(options, (res) => {
          let responseBody = '';
          
          res.on('data', (chunk) => {
            responseBody += chunk;
          });

          res.on('end', () => {
            try {
              if (res.statusCode && res.statusCode >= 400) {
                reject(new Error(`HTTP ${res.statusCode}: ${responseBody}`));
                return;
              }

              // Try to parse JSON response
              try {
                const jsonResponse = JSON.parse(responseBody);
                resolve(jsonResponse);
              } catch {
                // If not JSON, return text
                resolve({ response: responseBody });
              }
            } catch (error) {
              reject(error);
            }
          });
        });

        req.on('error', (error) => {
          reject(error);
        });

        // Write request body if present
        if (postData) {
          req.write(postData);
        }

        req.end();

      } catch (error) {
        console.error(`❌ Service call to ${serviceName} failed:`, (error as Error).message);
        reject(error);
      }
    });
  }

  /**
   * Validate user exists before creating order
   */
  async validateUser(userId: string): Promise<{ valid: boolean; user?: any }> {
    try {
      console.log(`🔍 Validating user ${userId} with user-service`);
      
      const userInfo = await this.makeServiceRequest(
        'user-service',
        `/customer/findcustomerbyid?customerid=${userId}`,
        'GET'
      );

      return {
        valid: true,
        user: userInfo
      };

    } catch (error) {
      console.error(`❌ User validation failed for ${userId}:`, (error as Error).message);
      return {
        valid: false
      };
    }
  }

  /**
   * Check if user is authenticated via login service
   */
  async validateAuthToken(token: string): Promise<{ valid: boolean; userId?: string }> {
    try {
      console.log(`🔐 Validating auth token with login-service`);
      
      const authResult = await this.makeServiceRequest(
        'login-service',
        '/auth/validate',
        'POST',
        { token }
      );

      return {
        valid: authResult.valid || false,
        userId: authResult.userId
      };

    } catch (error) {
      console.error(`❌ Auth token validation failed:`, (error as Error).message);
      return {
        valid: false
      };
    }
  }

  /**
   * Notify user service about order creation
   */
  async notifyOrderCreated(userId: string, orderId: string, orderData: any): Promise<void> {
    try {
      console.log(`📢 Notifying user-service about order ${orderId} for user ${userId}`);
      
      await this.makeServiceRequest(
        'user-service',
        '/notifications/order-created',
        'POST',
        {
          userId,
          orderId,
          orderData,
          timestamp: new Date().toISOString()
        }
      );

      console.log(`✅ Order notification sent successfully`);

    } catch (error) {
      // Don't fail the order creation if notification fails
      console.warn(`⚠️ Failed to notify user service about order:`, (error as Error).message);
    }
  }

  /**
   * Get user preferences for order processing
   */
  async getUserPreferences(userId: string): Promise<any> {
    try {
      console.log(`⚙️ Getting user preferences for ${userId}`);
      
      const preferences = await this.makeServiceRequest(
        'user-service',
        `/users/${userId}/preferences`,
        'GET'
      );

      return preferences;

    } catch (error) {
      console.warn(`⚠️ Failed to get user preferences:`, (error as Error).message);
      return {}; // Return empty preferences as fallback
    }
  }

  /**
   * Health check other services
   */
  async checkServiceHealth(serviceName: string): Promise<boolean> {
    try {
      await this.makeServiceRequest(serviceName, '/health', 'GET');
      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * Get all available services
   */
  async getAvailableServices(): Promise<string[]> {
    try {
      const services = ['user-service', 'login-service'];
      const availableServices: string[] = [];

      for (const service of services) {
        const isHealthy = await this.checkServiceHealth(service);
        if (isHealthy) {
          availableServices.push(service);
        }
      }

      return availableServices;
    } catch (error) {
      console.error('❌ Failed to check available services:', (error as Error).message);
      return [];
    }
  }
}

// Export singleton instance
const serviceClient = new ServiceClient();
export default serviceClient;